core/src/main/clojure/xtdb/log.clj

```diff
@@ -247,8 +250,56 @@
                        :kafka :xtdb.kafka/log)
                      opts))
 
-(defmethod ig/init-key :xtdb/log [_ ^Log$Factory factory]
-  (.openLog factory))
+(defmethod ig/prep-key :xtdb/log [_ factory]
+  {:block-cat (ig/ref :xtdb/block-catalog)
+   :factory factory})
+
+(defn tx-id->offset [^long tx-id]
+  (bit-and tx-id (dec (bit-shift-left 1 48))))
+
+(defn tx-id->epoch [^long tx-id]
+  (bit-shift-right tx-id 48))
+
+(defn class->log-type [log]
+  (let [class-name (str (class log))]
+    (cond
+      (str/includes? class-name "MemoryLog") :memory-log
+      (str/includes? class-name "LocalLog")  :local-directory-log
+      (str/includes? class-name "KafkaLog")  :kafka-log)))
+
+(def log-error-messages
+  {:memory-log
+   {:empty-log
+    "Starting node with a memory log and previously indexed values (latest completed offset: %s).\n\nIf this is intended and you wish to start from indexed data on storage, set the following in your config:\n\nlog: !InMemory\n  currentEpoch: %s\n"}
+
+   :local-directory-log
+   {:empty-log
+    "Starting node with an empty local directory log and previously indexed values (latest completed offset: %s).\n\nIf this is intended and you wish to start from indexed data on storage, set the following in your config to start a new epoch:\n\nlog: !Local\n  currentEpoch: %s\n"} 
+   
+   :kafka-log
+   {:empty-log
+    "Starting node with an empty Kafka log topic and previously indexed values (latest completed offset: %s).\n\nIf this is intended and you wish to start from indexed data on storage, set the following in your config to start a new epoch:\n\nlog: !Kafka\n  currentEpoch: %s\n"}})
+
+(defn ->empty-log-exception [log-type latest-offset ^long current-epoch]
+  (let [template (get-in log-error-messages [log-type :empty-log])]
+    (IllegalStateException.
+     (format template latest-offset (inc current-epoch)))))
+
+(defn validate-offsets [^Log log ^TransactionKey latest-completed-tx]
+  (when latest-completed-tx
+    (let [log-type (class->log-type log)
+          latest-completed-tx-id (.getTxId latest-completed-tx)
+          latest-offset (tx-id->offset latest-completed-tx-id)
+          latest-epoch (tx-id->epoch latest-completed-tx-id)
+          current-epoch (.getCurrentEpoch log)]
+      (when (= latest-epoch current-epoch)
```

Comment:
Could perhaps include a log when not validating offsets (ie, because epochs differ)